xonsh/environ.py

```diff
@@ -625,10 +639,14 @@ def __setitem__(self, key, val):
             val = ensurer.convert(val)
         self._d[key] = val
         self._detyped = None
+        if self.get('UPDATE_OS_ENVIRON'):
+            self.update_env()
```

Comment:
I think there is a more efficient way to do this.  Everytime a single key is set, the whole of `os.environ` is updated, however it should probably just replace that key.  I think the following would be better

```python
def __setitem__(self, key, val):
    ensurer = self.get_ensurer(key)
     if not ensurer.validate(val):
        val = ensurer.convert(val)
    self._d[key] = val
    self._detyped = None
    if self.get('UPDATE_OS_ENVIRON'):
        dval = ensurer.detype(val)
        os.environ[key] = dval
```